src/input/mpegts/linuxdvb/linuxdvb_ddci.c

```diff
@@ -675,28 +680,73 @@ linuxdvb_ddci_read_thread ( void *arg )
       }
       ddci_rd_thread->lddci_recv_pkgCntS += scrambled;
 
-      /* FIXME: split the received packets according to the PID in different
-       *        buffers and deliver them
-       * FIXME: How to determine the right service pointer?
-       */
-      /* as a first step we send the data to the associated service */
-      t = ddci_thread->lddci->t;
-      if (t) {
-        pthread_mutex_lock(&t->s_stream_mutex);
-        if (t->s_status == SERVICE_RUNNING) {
-          ts_recv_packet2((mpegts_service_t *)t, tsb, len);
-          ddci_rd_thread->lddci_recv_pkgCntW += num_pkg;
+      len2 = len;
+      while (len2 > 0) {
+        uint16_t pid;
+        int llen;
+
+        /* Send all special PIDs (<20: PAT, EIT,..) to all services
+         * To do that, we need first split the stream in junks with the same
+         * PID
+         */
+
+        /*
+         * mask
+         *  0 - 0xFF - sync word 0x47
+         *  1 - 0x80 - transport error
+         *  1 - 0x1F - pid high
+         *  2 - 0xFF - pid low
+         *  3 - 0x00 - ignore scrambled flags
+         */
+        llen = mpegts_word_count(tsb, len2, 0xFF9FFF00);
+
+        pid = (tsb[1] << 8) | tsb[2];
+        pid &= 0x1FFF;
+
+        /* All PIDs < 0x20 need to be sent to all services */
+        if (pid < 0x20) {
```

Comment:
This is already handled in mpegts_input_postdemux() - the shared PIDs are marked as MPS_STREAM. I would propose to add the CAT pid filter to this fcn and remove completely this loop for MCD. I would propose to introduce a new MPS_NOPOSTDEMUX bit flag (src/descrambler/dvbcam.c - look for DVB_CAT_PID) and skip PIDs marked with this flag in mpegts_input_postdemux().